claat/serve.go

```diff
@@ -0,0 +1,217 @@
+`
+	err := os.MkdirAll(depsDir, 0755)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	// Go get the dependencies.
+	err = fetchRepo(depsDir, "googlecodelabs/codelab-components#^1.0.0")
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	//os.Rename(depsDir+"/codelab-components", depsDir+"/google-codelab-elements")
+	//os.Rename(depsDir+"/code-prettify", depsDir+"/google-prettify")
+	err = os.MkdirAll(elemDir, 0755)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	if _, err := os.Stat(elemFile); os.IsNotExist(err) {
+		f, err := os.Create(elemFile)
+		if err != nil {
+			log.Fatal("Cannot create file", err)
+		}
+		defer f.Close()
+		fmt.Fprintf(f, codelabElem)
+	}
+	fmt.Println("Dependencies installed.")
+
+	port := "9090"
+	http.Handle("/", http.FileServer(http.Dir(".")))
+	fmt.Println("Serving on localhost:" + port + ", opening browser tab now...")
+	openBrowser("http://127.0.0.1:" + port)
+	err = http.ListenAndServe(":"+port, nil) // set listen port
+	if err != nil {
+		log.Fatal("ListenAndServe: ", err)
+	}
+}
+
+// downloadFile will download a url to a local file. It's efficient because it will
+// write as it downloads and not load the whole file into memory.
+func downloadFile(filepath string, url string) error {
+	out, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func fetchRepo(depsDir string, repoVers string) error {
+	// parse url into repoName and version
+	s := strings.Split(repoVers, "#^")
+	vers := ""
+	if len(s) > 1 {
+		vers = s[1]
+	}
+	s = strings.Split(s[0], "/")
+	user := s[0]
+	repo := s[1]
+
+	// if repo already exists locally, return immediately, we're done.
+	if _, err := os.Stat(depsDir + "/" + repo); os.IsExist(err) {
+		return nil
+	}
+	url := "https://github.com/" + user + "/" + repo + "/archive/" + vers + ".zip"
+	zipFile := depsDir + "/" + repo + ".zip"
+	fmt.Println("Downloading " + url)
+	err := downloadFile(zipFile, url)
+	if err != nil {
+		return err
+	}
+	err = unzip(zipFile, depsDir)
+	if err != nil {
+		return err
+	}
+	os.Remove(zipFile)
+	os.Rename(depsDir+"/"+repo+"-"+vers, depsDir+"/"+repo)
+
+	// if unzipped archive contains a bower.json, parse it, and for each dependency therein,
+	// recursively fetch the corresponding repo.
+	bowerFile := depsDir + "/" + repo + "/bower.json"
+	if _, err := os.Stat(bowerFile); os.IsNotExist(err) {
+		return nil
+	}
+	raw, err := ioutil.ReadFile(bowerFile)
+	if err != nil {
+		return err
+	}
+	type Dep struct {
+		Name     string
+		RepoVers string
+	}
+	var objmap map[string]*json.RawMessage
+	err = json.Unmarshal(raw, &objmap)
+	if err != nil {
+		return err
+	}
+	var deps map[string]string
+	err = json.Unmarshal(*objmap["dependencies"], &deps)
+	if err != nil {
+		return err
+	}
+	for _, v := range deps {
+		err = fetchRepo(depsDir, v)
```

Comment:
Instead of unmarshalling twice, you could have:

```go
type bower struct {
  Dependencies map[string]string
}
// Unmarshal once ...
for dest, comp := range b.Dependencies {
  fetchRepo(basedir + dest, comp)
}
```

Also, note the "basedir +dest" correction. The key specifies the name of the local dir.

Reply:
I think you don't even need to define a type, use anonymous:

```go
var bower struct {
  Dependencies map[string]string
}
```

Reply:
Looks a lot like your prev example. Did you mean "var struct" minus the struct tag?

Reply:
I get it, you're saying I can use the bower type without naming an instance.

Reply:
fixed.